traffic_router/core/src/main/java/com/comcast/cdn/traffic_control/traffic_router/core/router/TrafficRouter.java

```diff
@@ -559,9 +563,47 @@ public HTTPRouteResult multiRoute(final HTTPRequest request, final Track track)
 		return routeResult;
 	}
 
-	public String buildPatternBasedHashString(final DeliveryService deliveryService, final String requestPath) {
+	/**
+	 * Creates a string to be used in consistent hashing
+	 *
+	 * This uses simply the request path by default, but will consider any and all Query Parameters
+	 * that are in deliveryService's {@link DeliveryService.consistentHashQueryParams} set as well.
+	 * It will also fall back on the request path if the query parameters are not UTF-8-encoded.
+	 *
+	 * @param deliveryService The Delivery Service being requested
+	 * @param request An {@link HTTPRequest} representing the client's request.
+	 * @return A string appropriate to use for consistent hashing to service the request
+	*/
+	@SuppressWarnings({"PMD.CyclomaticComplexity"})
+	public String buildPatternBasedHashString(final DeliveryService deliveryService, final HTTPRequest request) {
+		final String requestPath = request.getPath();
 		if (deliveryService.getConsistentHashRegex() != null && !deliveryService.getConsistentHashRegex().isEmpty() && !requestPath.isEmpty()) {
-			return buildPatternBasedHashString(deliveryService.getConsistentHashRegex(), requestPath);
+			final StringBuilder hashString = new StringBuilder(request.getPath());
+			final SortedSet<String> qparams = new TreeSet<String>();
+			try {
+				for (final String qparam : request.getQueryString().split("&")) {
+					if (qparam.isEmpty()) {
+						continue;
+					}
+
+					String[] parts = qparam.split("=");
+					for (short i = 0; i<parts.length; ++i) {
+						parts[i] = URLDecoder.decode(parts[i]);
+					}
+
+					if (deliveryService.getConsistentHashQueryParams().contains(parts[0])) {
+						qparams.add(String.join("=", parts));
+					}
+				}
+
+				for (final String qparam : qparams) {
+					hashString.append(qparam);
+				}
+			} catch (Exception e) {
+				return requestPath;
+			}
+
+			return buildPatternBasedHashString(deliveryService.getConsistentHashRegex(), hashString.toString());
```

Comment:
So I think the pattern-based consistent hashing feature was meant to only be for the request path, but now we're adding on the consistent hash query parameters. Is that going to be a problem? Should we build the pattern-based consistent hash string first (using just the path), _then_ append any consistent hash query params onto the path?

Reply:
I thought it'd be fine to do the regex on the whole path+query params. I don't _think_ it would bother things only meant to match on paths  - unless maybe you have a path in a param value? 

It sort of expands the functionality of the regex consistent hashing, which I thought would be neat - like `.*\?test=\d+.*` to consistent hash based on the `test` param but only if it's a number or something.

Idk, do *you* think it'll cause problems? We haven't issued a release with CHR in it, so idk if breaking that behavior is a problem in itself.

Reply:
I don't think we should expand the functionality of the regex consistent hashing to apply to the consistent hash query params (CHQPs). If we can keep the two features separate, they're easier to reason about. With CHQPs we're saying "these are the query parameters that change the content of a request path". Since those query parameters produce different content for the same path, we should hash them differently. I'm not sure we should take that any further by being able to configure something like "only consider this query param if its value matches this pattern". Trying to come up with a single regex that extracts a hash string from the path *and* the qstring seems overly complicated.

It just seems like keeping the two features separate for now would make them both easier to use on their own. If we have a requirement in the future to only consider certain query parameters if they match a pattern, then I'd think each qparam that had to match a pattern would get its own regex instead of trying to configure patterns for all of those qparams in the one consistent hash regex.

Reply:
If you say so ¯\_(ツ)_/¯. Done (pending branch merge).